Unicode/makeutype.c

```diff
@@ -630,6 +683,178 @@ static void dumparabicdata(FILE *header) {
     fclose( data );
 }
 
+static void buildtables(FILE *data,long unsigned int *dt,int s,int m,char *t) {
+/* Build ligature/vulgar/fraction table as a condensed uint16 array for the */
+/* range of 't'16{0...s} and then as uint32 for the range 't'32{(s+1)...m}. */
+    int i,j;
+
+    fprintf( data, "const uint16 %s16[] = {\n", t );
+    for ( i=j=0; i<s; i=i+j ) {
+	fprintf( data, "  0x%04x", dt[i] );
+	for ( j=1; j<8 && i+j<s; ++j ) {
+	    fprintf(data, ", 0x%04x", dt[i+j] );
+	}
+	if ( i+j>=s )
+	    fprintf(data, "};\n\n" );
+	else
+	    fprintf(data, ",\n" );
+    }
+    if ( s<m ) {
+	fprintf( data, "const uint32 %s32[] = {\n", t );
+	for ( i=s; i<m; i=i+j ) {
+	    fprintf( data, "  0x%08x", dt[i] );
+	    for ( j=1; j<4 && i+j<m; ++j ) {
+		fprintf(data, ", 0x%08x", dt[i+j] );
+	    }
+	    if ( i+j>=m )
+		fprintf(data, "};\n\n" );
+	    else
+		fprintf(data, ",\n" );
+	}
+    }
+}
+
+static void dumpbsearch(FILE *data,int s) {
+/* Customize bsearch sizes to match data table sizes for uint16 and uint32. */
+    fprintf( data, "static int compare_codepoints%d(const void *uCode1, const void *uCode2) {\n", s );
+    fprintf( data, "    const uint%d *cp1 = (const uint%d *)(uCode1);\n", s, s );
+    fprintf( data, "    const uint%d *cp2 = (const uint%d *)(uCode2);\n", s, s );
+    fprintf( data, "    return( (*cp1 < *cp2) ? -1 : ((*cp1 == *cp2) ? 0 : 1) );\n}\n\n" );
+}
+
+static void dump_getU(FILE *data,long unsigned int *dt,int s,int m,char *t, char *n) {
+/* Get the 'n'th unicode value in the table, where {0<=n<=(sizeof(table)-1} */
+/* Treat both (uint16, uint32) tables as if it is only one table. Error=-1. */
+    fprintf( data, "int32 %s_get_U(int n) {\n", n );
+    fprintf( data, "    if ( n<0 || n>=%d )\n\treturn( -1 );\n    ", m );
+    if ( s<m )
+	fprintf( data, "if ( n<%d )\n\t", s );
+    fprintf( data, "return( (int32)(%s16[n]) );\n", t );
+    if ( s<m )
+    fprintf( data, "    else\n\treturn( (int32)(%s32[n-%d]) );\n", t, s );
+    fprintf( data, "}\n\n" );
+}
```

Comment:
We probably ought to rename parameter n to something like nam so as to avoid confusion with the other n.

Are these comments right?
```
// m is the maximum size of the table.
// s is the size of the lower partition of the table.
// t is the first part of the data type name.
// nam is the function name base.
```